app/queries/articles/feeds/filter_query.rb

```diff
@@ -0,0 +1,92 @@
+module Articles
+  module Feeds
+    class FilterQuery
+      # [Ridhwana]: Important to note that this is a POC and is not thoroughly tested manually or
+      # in an automated fashion as yet. More verification is required.
+      MINIMUM_SCORE = -20
+
+      def self.call(...)
+        new(...).call
+      end
+
+      def initialize(feed_type: "explore", minimum_score: MINIMUM_SCORE,
+                     number_of_articles: ::Article::DEFAULT_FEED_PAGINATION_WINDOW_SIZE,
+                     page: 1, timeframe: nil, tag: nil, user: nil)
+        @filtered_articles = Article.all
+        @feed_type = feed_type
+        @minimum_score = minimum_score
+        @number_of_articles = number_of_articles
+        @page = page
+        @timeframe = timeframe
+        @tag = tag
+        @user = user
+      end
+
+      def call
+        if @tag.present?
+          @filtered_articles = filter_by_tags
+        end
+
+        if @feed_type == "following"
+          @filtered_articles = filter_by_following_tags
+        end
+
+        @filtered_articles = base_operations
+
+        if @timeframe == ::Timeframe::LATEST_TIMEFRAME
+          @filtered_articles = latest_feed
+        elsif @timeframe.in?(::Timeframe::FILTER_TIMEFRAMES)
+          @filtered_articles = timeframe_feed
+        end
+
+        @filtered_articles = filter_out_hidden_tagged_articles
+
+        @filtered_articles
+      end
+
+      private
+
+      def base_operations
+        @filtered_articles
+          .published
+          .limited_column_select
+          .includes(top_comments: :user)
+          .includes(:distinct_reaction_categories)
+          .page(@page)
+          .per(@number_of_articles)
+      end
+
+      def filter_by_following_tags
+        return unless (followed_tags = @user.cached_followed_tag_names).any?
```

Comment:
I think there's a bug here that suggests we _might_ want to try a subtly different approach to writing this class.

So: the bug comes from returning `nil` from the guard clause here and then assigning that to `@filtered_articles` on line 31. (There's a similar bug around line 74 in `filter_out_hidden_tagged_articles`.)

Approach: Basically, we're _going_ to need to have guard clauses around these filters. As I see it, I think we have two options:

* _Every_ `filter_whatever` method _*always*_ returns the modified scope — if we want to apply them conditionally, we _must_ check the condition _before_ calling `filter_whatever`.
* _Every_ `filter_whatever` applies the scope modifications _directly_ to `@filtered_articles` — if we want to apply filters conditionally, we can `return` early to skip applying filters _or_ apply the condition before calling `filter_whatever`.

(I mean, _technically_, mixing & matching these _can work_ but it seems to me that we're likely to have more problems from that approach.)

Personally, I would generally prefer the latter option (as the flexibility seems less likely to cause problems). I suspect there's a "styleguide" somewhere that will warn us against the second option ("modifying state as a side-effect") but I would argue here that this class's _main effect_ is applying filters to a scope.

As a sort-of bonus, if we _did_ move _all_ of the conditionals into the `filter_whatever` methods, call becomes quite tidy:

```ruby
def call
  basic_filters
  filter_by_selected_tags
  filter_by_following_tags
  latest_feed || timeframe_feed
  filter_out_hidden_tag_articles
  filtered_articles
end
```